nautobot/dcim/forms.py

```diff
@@ -430,6 +433,28 @@ class Meta:
             "outer_unit": StaticSelect2(),
         }
 
+    def clean(self):
+        cleaned_data = self.cleaned_data
+        site = cleaned_data.get("site")
+
+        if self.instance:
+            # If the site is changed, the rack post save signal attempts to update the rack devices,
+            # which may result in an Exception.
+            # To avoid an unhandled exception in signal, catch this error here.
+            duplicate_devices_names = (
+                Device.objects.values_list("name", flat=True)
+                .annotate(name_count=Count("name"))
+                .filter(name_count__gt=1)
+            )
+            duplicate_devices = Device.objects.filter(site=site, name__in=list(duplicate_devices_names)).values_list(
+                "name", flat=True
+            )
+            if duplicate_devices:
+                raise ValidationError(
+                    {"site": f"Device with `name` in {list(duplicate_devices)} and site={site} already exists."}
+                )
+        return cleaned_data
+
```

Comment:
Site-specific code - IIRC Timizuo said there's no equivalent problem with `location`.
```suggestion
```

Reply:
This answers my question below. @timizuoebideri1 Why do we not have validation problems when changing a Rack's location?

Reply:
@glennmatthews @timizuoebideri1 Can I get an update to the ☝️?

Why don't we have the same problem changing Rack location?

I think Device.location is still part of a device's uniqueness constraint?

Reply:
Just tested it. We *do* get an `IntegrityError` on the console with the Location change of a Rack:

```
DETAIL:  Key (location_id, tenant_id, name)=(9941a4d8-9952-4513-a43a-a0bc6e5094de, 7fb7fa83-f713-4df8-b8c6-f57c2bf2680c, example) already exists.
Traceback (most recent call last):
  File "/opt/nautobot/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/opt/nautobot/lib/python3.8/site-packages/django_prometheus/db/common.py", line 71, in execute
    return super().execute(*args, **kwargs)
psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "dcim_device_location_id_tenant_id_name_2259bd02_uniq"
DETAIL:  Key (location_id, tenant_id, name)=(9941a4d8-9952-4513-a43a-a0bc6e5094de, 7fb7fa83-f713-4df8-b8c6-f57c2bf2680c, example) already exists.


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/source/nautobot/core/views/generic.py", line 417, in post
    obj = form.save()
  File "/source/nautobot/extras/forms/mixins.py", line 632, in save
    obj = super().save(commit)
  File "/opt/nautobot/lib/python3.8/site-packages/django/forms/models.py", line 468, in save
    self.instance.save()
  File "/opt/nautobot/lib/python3.8/site-packages/django/db/models/base.py", line 739, in save
    self.save_base(using=using, force_insert=force_insert,
  File "/opt/nautobot/lib/python3.8/site-packages/django/db/models/base.py", line 787, in save_base
    post_save.send(
  File "/opt/nautobot/lib/python3.8/site-packages/django/dispatch/dispatcher.py", line 180, in send
    return [
  File "/opt/nautobot/lib/python3.8/site-packages/django/dispatch/dispatcher.py", line 181, in <listcomp>
    (receiver, receiver(signal=self, sender=sender, **named))
  File "/source/nautobot/dcim/signals.py", line 145, in handle_rack_location_change
    device.save()
  File "/source/nautobot/dcim/models/devices.py", line 680, in save
    super().save(*args, **kwargs)
  File "/opt/nautobot/lib/python3.8/site-packages/django/db/models/base.py", line 739, in save
    self.save_base(using=using, force_insert=force_insert,
  File "/opt/nautobot/lib/python3.8/site-packages/django/db/models/base.py", line 776, in save_base
    updated = self._save_table(
  File "/opt/nautobot/lib/python3.8/site-packages/django/db/models/base.py", line 858, in _save_table
    updated = self._do_update(base_qs, using, pk_val, values, update_fields,
  File "/opt/nautobot/lib/python3.8/site-packages/django/db/models/base.py", line 912, in _do_update
    return filtered._update(values) > 0
  File "/opt/nautobot/lib/python3.8/site-packages/django/db/models/query.py", line 802, in _update
    return query.get_compiler(self.db).execute_sql(CURSOR)
  File "/opt/nautobot/lib/python3.8/site-packages/django/db/models/sql/compiler.py", line 1559, in execute_sql
    cursor = super().execute_sql(result_type)
  File "/opt/nautobot/lib/python3.8/site-packages/django/db/models/sql/compiler.py", line 1175, in execute_sql
    cursor.execute(sql, params)
  File "/opt/nautobot/lib/python3.8/site-packages/debug_toolbar/panels/sql/tracking.py", line 247, in execute
    return self._record(self.cursor.execute, sql, params)
  File "/opt/nautobot/lib/python3.8/site-packages/debug_toolbar/panels/sql/tracking.py", line 168, in _record
    return method(sql, params)
  File "/opt/nautobot/lib/python3.8/site-packages/django/db/backends/utils.py", line 98, in execute
    return super().execute(sql, params)
  File "/opt/nautobot/lib/python3.8/site-packages/django/db/backends/utils.py", line 66, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/opt/nautobot/lib/python3.8/site-packages/django/db/backends/utils.py", line 75, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/opt/nautobot/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/opt/nautobot/lib/python3.8/site-packages/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/opt/nautobot/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/opt/nautobot/lib/python3.8/site-packages/django_prometheus/db/common.py", line 71, in execute
    return super().execute(*args, **kwargs)
django.db.utils.IntegrityError: duplicate key value violates unique constraint "dcim_device_location_id_tenant_id_name_2259bd02_uniq"
DETAIL:  Key (location_id, tenant_id, name)=(9941a4d8-9952-4513-a43a-a0bc6e5094de, 7fb7fa83-f713-4df8-b8c6-f57c2bf2680c, example) already exists.
```

but in `next` we have [(temporary?) code in ObjectEditView that catches IntegrityErrors and tries to guess what the problem is to report to the user](https://github.com/nautobot/nautobot/blob/next/nautobot/core/views/generic.py#L441-L483), which is what results in the UI reporting `Rack with this Rack Group, Name already exists`, which is incorrect and misleading. We probably need to chase this down as a new issue but I don't think it should block this PR.

Reply:
Doing some local investigation, I think we *should* be able to revert that temporary code with some workarounds elsewhere.

Reply:
See #4115.

Reply:
```suggestion
    def clean(self):
        cleaned_data = self.cleaned_data
        location = cleaned_data.get("location")

        if self.instance:
            # If the location is changed, the rack post save signal attempts to update the rack devices,
            # which may result in an Exception.
            # To avoid an unhandled exception in signal, catch this error here.
            duplicate_devices_names = (
                Device.objects.values_list("name", flat=True)
                .annotate(name_count=Count("name"))
                .filter(name_count__gt=1)
            )
            duplicate_devices = Device.objects.filter(location=location, name__in=list(duplicate_devices_names)).values_list(
                "name", flat=True
            )
            if duplicate_devices:
                raise ValidationError(
                    {"location": f"Device with `name` in {list(duplicate_devices)} and location={location} already exists."}
                )
        return cleaned_data

```